### PR TITLE
feat(rust-port): Split up shim and cli

### PR DESCRIPTION
**Pull request by NicholasLYang**
We split up the shim and CLI into two files. The shim runs a barebones parser that just parses out `--cwd`, `--single-package`, and `--skip-infer` flags. It does the package inference from these flags, then either calls into the current binary or executes local turbo.

The goal of this split is to allow global turbo to maintain forward compatibility with local turbo. If local turbo introduces a new flag, global turbo should allow it.

This does come with some drawbacks:

- All commands require being in a valid repo. That means that commands like `link`, `unlink`, `login` might break for users who execute them in invalid repos.
- We also need to make --single-package a global flag as we cannot determine if we are running a command when we add the flag.

### CODE REVIEW

File: crates/turborepo/tests/mod.rs [hunk 1]
```diff
@@ -0,0 +1,58 @@
+use assert_cmd::Command;
```
Comment (chris-olszewski) [inline crates/turborepo/tests/mod.rs:1]: This file should be deleted since we won't be running these tests in CI and all of the functionality is covered by integration tests.

File: crates/turborepo/Cargo.toml [hunk 2]
```diff
@@ -6,22 +6,22 @@ license = "MPL-2.0"
...
 clap = { version = "4.0.22", features = ["derive"] }
 clap_complete = "4.0.6"
 predicates = "2.1.1"
 serde = { version = "1.0.145", features = ["derive"] }
 serde_json = "1.0.86"
 serde_yaml = "0.8.26"
-turbo-updater = { path = "../crates/turbo-updater" }
-tiny-gradient      = "0.1"
+tiny-gradient = "0.1"
+turbo-updater = { path = "../turbo-updater" }
```
Comment (chris-olszewski) [inline crates/turborepo/Cargo.toml:26]: `turborepo-lib` pulls this in now
```suggestion
```

File: docs/components/clients/users.ts [hunk 3]
```diff
@@ -524,6 +524,6 @@ export const users: Array<TurboUser> = [
     pinned: true,
     style: {
       width: 150,
-    }
+    },
```
Comment (chris-olszewski) [inline docs/components/clients/users.ts:527]: Unrelated change?

File: crates/turborepo-lib/src/shim.rs [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+pub fn run() -> Result<Payload> {
+    let args = ShimArgs::parse()?;
+
+    if args.skip_infer || args.has_help_flag || is_turbo_binary_path_set() {
+        return cli::run();
+    }
+
+    let repo_state = RepoState::infer(&args.cwd)?;
+
+    repo_state.run_correct_turbo(args)
```
Comment (chris-olszewski) [inline crates/turborepo-lib/src/shim.rs:259]: I think we should be executing current binary if we can't infer the repo. This will allow `--version` `--help` and `login` etc to work outside of a repo.
```suggestion
    if args.skip_infer || is_turbo_binary_path_set() {
        return cli::run();
    }

    if let Ok(repo_state) = RepoState::infer(&args.cwd) {
        repo_state.run_correct_turbo(args)
    }

    cli::run()
```

**Review by chris-olszewski — CHANGES_REQUESTED** [review]

File: crates/turborepo-lib/src/shim.rs [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+    }
+
+    fn spawn_local_turbo(&self, local_turbo_path: &Path, mut shim_args: ShimArgs) -> Result<i32> {
+        let cwd = self
+            .root
+            .to_str()
+            .ok_or_else(|| anyhow!("Root directory path is invalid unicode"))?
+            .to_string();
+
+        let mut raw_args: Vec<_> = vec!["--skip-infer".to_string(), "--cwd".to_string(), cwd];
```
Comment (chris-olszewski) [inline crates/turborepo-lib/src/shim.rs:218]: Shouldn't we be setting the new process's cwd instead of using the flag since it was added semi-recently?

**Comment by chris-olszewski** [comment]
We can get rid of the Go `RunPayload` structure now, correct?

File: docs/components/clients/users.ts (hunk 3 above)
Comment (NicholasLYang) [reply docs/components/clients/users.ts:527]: yeah although this is a prettier fix which should be triggered by the pre-commit hook. Probably should merge it in because it'll show up on any other PR

File: crates/turborepo-lib/src/shim.rs [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+        let args = env::args().skip(1);
+        for arg in args {
+            // We've seen a `--` and therefore we do no parsing
+            if is_forwarded_args {
+                remaining_args.push(arg);
+            } else if arg == "--help" || arg == "-h" {
+                has_help_flag = true;
+            } else if arg == "--skip-infer" {
+                skip_infer = true;
+            } else if arg == "--single-package" {
```
[the remark this answers is not in the document]
Comment (NicholasLYang) [reply crates/turborepo-lib/src/shim.rs:41]: We'd need to do the `contains` check on the non-forwarded args so we'd have to split those up.

**Comment by NicholasLYang** [comment]
I don't think we can. It's what stores the arguments to a run command, no? Or are you thinking of a different struct?

File: crates/turborepo-lib/src/shim.rs (hunk 5 above)
Comment (NicholasLYang) [reply crates/turborepo-lib/src/shim.rs:213]: Valid. I was not sure if `--cwd` did something special in Go-land

File: crates/turborepo-lib/src/shim.rs [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+        let mut found_cwd_flag = false;
+        let mut cwd: Option<PathBuf> = None;
+        let mut skip_infer = false;
+        // We check for --single-package so that we don't add it twice
+        let mut single_package = false;
+        let mut remaining_args = Vec::new();
+        let mut is_forwarded_args = false;
+        let mut has_help_flag = false;
+        let args = env::args().skip(1);
+        for arg in args {
```
[the remark this answers is not in the document]
Comment (NicholasLYang) [reply crates/turborepo-lib/src/shim.rs:35]: Do we want to reject commands that pass multiple `--cwd`? Or are you saying that we should verify it but still keep the same behavior?

**Comment by NicholasLYang** [comment]
@chris-olszewski the only part left is potential refactoring of the basic parser. Personally I don't think it's that hard to read, considering it's a fairly straightforward loop with different conditions. But that's just me.

File: examples/non-monorepo/yarn.lock [hunk 8]
```diff
@@ -1,1815 +0,0 @@
-# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (chris-olszewski) [inline examples/non-monorepo/yarn.lock:1]: Was this needed?

**Review by chris-olszewski — APPROVED** [review]
We can keep the shim parsing as is since you're confident in it.

File: examples/non-monorepo/yarn.lock (hunk 8 above)
Comment (NicholasLYang) [reply examples/non-monorepo/yarn.lock:1]: Probably not although it's not necessary either way

**Review by tknickman — APPROVED** [review]
